CEECORE-912, update electrode-hapi-compat to be aware of @hapi/hapi 18

scrajesh opened the pull request:
- made changes to account for @hapi/hapi package

- Following new apis exposed for hapi@18
   - isHapi18() tells(true/false) if the hapi version used is v18
   - _testSetHapi18() used to modify the above flag during writing tests for hapi@18

- hapi@18 follows the same steps to register a hapi plugin
   - universalHapiPlugin() recognizes v18 similar to v17 and registers the plugin

> The unit tests mostly defaults to v16, it's the version specified in the package.json
    -- whenever this version is changed, we might need to make appropriate unit test changes.
    -- this PR only includes new tests for v18.

scrajesh committed:
CEECORE-912, update electrode-hapi-compat to be aware of @hapi/hapi 18

jchip commented on lib/util.js:
```diff
@@ -1,40 +1,58 @@
+    hapiPkg = require("@hapi/hapi/package");
   } catch (err) {
-    // ignore
+    hapiPkg = require("hapi/package");
```
this need a try catch also

jchip commented on lib/is-valid-object.js:
```diff
@@ -0,0 +1,10 @@
+
+function isValidObject(obj) {
+  if (typeof obj === "object" && obj !== null) {
+    return Object.keys(obj).length;
```
return a boolean

jchip commented on lib/util.js:
```diff
@@ -1,40 +1,58 @@
   return _isHapi17;
 }
 
+function isHapi18() {
```
let's name this `isHapi18OrUp`, and check for `>= 18`

christianlent commented on lib/util.js:
```diff
@@ -1,40 +1,58 @@
-let _isHapi17 = hapiMajor() >= HAPI17;
+
+const hapiV = hapiMajor();
+let _isHapi17 = hapiV >= HAPI17;
```
Should the getter for this also be named "isHapi17OrUp"? Or should this be a strict equality check?

jchip replied:
we can add `isHapi17OrUp` but can't change the existing `isHapi17`.
`isHapi17` should do `>=`, so existing modules will work with 18 w/o change, except an update on this module.

scrajesh replied:
Had the same doubt and i wanna leave it for discussion. I prefer strict equality check though.

christianlent replied:
It seems clear to me based on Joel's comment that you did the right thing - folks are using the isHapi17 function as if it's isHapi17OrUp I suppose? As Joel said, you could optionally add a isHapi17OrUp function, but it would essentially have an identical implementation.

datvong-wm commented on lib/util.js:
```diff
@@ -1,40 +1,58 @@
+  if (_isHapi17 || _isHapi18) {
     return {
-      register: registers.hapi17,
+      register: _isHapi18 ? registers.hapi18 : registers.hapi17,
```
are we expecting users to specify hapi17, hapi18, and hapi16?

jchip replied:
`hapi17` needs to continue to work for both 17, 18, 19, 20, ...., until they change and break the plugin interface, but we can add a new field `hapi17OrUp`.

jchip replied:
please remove `hapi18`

scrajesh replied:
I want to specify only for hapi16 and if 17 and above it has to be simply `registers.hapi`, kinda default. Want to know all the thoughts on this..

jchip replied:
that's actually a good idea, except it would not look very clear on user's side:

```js
module.exports = universalHapiPlugin({
  hapi16: legacyPlugin,
  hapi: otherPlugin
})
```

vs: 

```js
module.exports = universalHapiPlugin({
  hapi16: legacyPlugin,
  hapi17OrUp: otherPlugin
})
```

jchip replied:
in any case, `hapi17` needs to continue to work, else it's a breaking change.

scrajesh replied:
@jchip, do you want me to replace registers.hapi18 with registers. hapi17OrUp ?
Or simply should i remove .hapi18 and use the current .hapi17 for the plugins ?

scrajesh committed:
PR feedback

- isValidObject, modified to return boolean instead of number of keys
- tryRequire, new method that imports a file withing try-catch block; will use this instead of nesting up try-catch block.
- isHapi18() renamed to isHapi18OrUp()
- plugin registers, both hapi17 and hapi17OrUp works with hapi@17/18

jchip commented on lib/try-require.js:
```diff
@@ -0,0 +1,13 @@
+    fileContent = require(file);
+  } catch (err) {
+    // ignore
+  }
```
nit:
```js
try {
  return require(file);
} catch(err) {
  return undefined;
}
```

scrajesh replied:
updated

jchip reviewed (approved):
LGTM with a nit

scrajesh committed:
return undefined explicitly if an issue with importing the file
